### PR TITLE
Use JS regex syntax

### LINKED ISSUE #607: Support for ECMA 262 regex patterns
The value of [the `pattern`] keyword [SHOULD be a valid regular expression](https://json-schema.org/latest/json-schema-validation.html#pattern), according to the ECMA 262 regular expression dialect.  However [sticking to a common subset of the syntax is recommended](https://json-schema.org/understanding-json-schema/reference/regular_expressions.html).

I spent the weekend doing some research on the differences between Python and JS regex, and seeing how difficult it might be to patch over them.  In short: https://pypi.org/project/js-regex/ now exists. (🎉)

It currently supports Python 3.6+ only, but solely because that's a much nicer way to develop a new project - if you were willing to use it in `jsonschema` I'd be happy to make a back-compatible release to support Pythons 2.7 and 3.5.

### PR DESCRIPTION
**Pull request by Zac-HD**
Note that `js_regex.compile` has a cache, just like `re.compile`, so calling it in loops does a lot of dict lookups but very little work.

Closes #607.

### CODE REVIEW

File: jsonschema/_utils.py [hunk 1]
```diff
@@ -92,10 +93,10 @@ def find_additional_properties(instance, schema):
     """
 
     properties = schema.get("properties", {})
-    patterns = "|".join(schema.get("patternProperties", {}))
+    patterns = "|".join(sorted(schema.get("patternProperties", {})))
```
Comment (Zac-HD) [inline jsonschema/_utils.py]: Sorting ensures that this is always a single entry in the regex cache, which is quite important for performance (of this tiny function, anyway).

**Comment by Julian** [comment]
Hooray, thanks!

Can you also enable the tests for this in the test suite? Should basically just be adding a line like this: https://github.com/Julian/jsonschema/blob/master/jsonschema/tests/test_jsonschema_test_suite.py#L71 to each of the drafts.

**Comment by Zac-HD** [comment]
Done!  I've left out Draft3 because it doesn't have the relevant test cases upstream, which I think is json-schema-org/JSON-Schema-Test-Suite#274.  

But drafts 4, 6, and 7 should all be validated... and there's a couple of failures.  Welp.   ...I think I'll need to take a little longer to debug that, and ensure that I got the everything right upstream :disappointed:

**Comment by Zac-HD** [comment]
@Julian - I think this just need to be run on top of the *fixed* version of my upstream tests, and we'll be good to go :smile:

**Comment by Julian** [comment]
Think we're getting there!

I think there's still one that looks like it's failing: https://travis-ci.org/Julian/jsonschema/jobs/594558002#L3575

**Comment by Zac-HD** [comment]
Aaaah, I see what's happening there: this *isn't* one of the "does it match" tests I added - it's actually for `{type: string, format: regex}`.  Compiling with `js_regex` instead of `re` should do the trick :smile:

**Comment by Zac-HD** [comment]
And done! :sparkles::tada::rocket:

**Comment by Julian** [comment]
Awesome, well done man, really appreciated.

**Comment by Zac-HD** [comment]
@Julian - any advice on when this will make it to PyPI?  I've been asked to support it in `hypothesis-jsonschema`, but doing so before `jsonschema` itself will cause lots of test failures... 

If it's going to be a while I can write a shim, but I'd feel pretty silly if I did that and you updated the day after!

**Comment by Julian** [comment]
Probably today :)

**Comment by Zac-HD** [comment]
Awesome!  I'll aim to do my integration tomorrow then :grin:

**Comment by Julian** [comment]
(Done) :)

**Pull request merged**